src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4374,15 +4530,15 @@ public override BoundNode VisitAsOperator(BoundAsOperator node)
                             if (operandType?.IsValueType == true)
                             {
                                 // https://github.com/dotnet/roslyn/issues/29959 We currently don't worry about a pathological case of boxing nullable value known to be not null
-                                nullableAnnotation = operandType.IsNullableType() ?  NullableAnnotation.Nullable : NullableAnnotation.NotNullable;
+                                nullableAnnotation = (operandType.IsNullableType() && _resultType.NullableAnnotation.IsAnyNullable()) ? NullableAnnotation.Nullable : NullableAnnotation.NotNullable;
```

Comment:
>_resultType [](start = 86, length = 11)

Since we are starting using ```_resultType```, I think it would be better to explicitly call ```VisitRvalue(node.Operand);``` at the top of the method instead of ```var result = base.VisitAsOperator(node);```. This way there will be no confusion what type does ```_resultType``` represent. Also, it would be good to remove ```operandType``` and replace its usage with ```_resultType```. This way it will be no confusion, or risk to combine checks for different types together.